editor/import/resource_importer_scene.cpp

```diff
@@ -302,159 +262,89 @@ Node *ResourceImporterScene::_fix_node(Node *p_node, Node *p_root, Map<Ref<Mesh>
 				}
 
-				if (_teststr(name, "colonly")) {
-					fixed_name = _fixstr(name, "colonly");
-				} else if (_teststr(name, "convcolonly")) {
-					fixed_name = _fixstr(name, "convcolonly");
-				}
-
-				ERR_FAIL_COND_V(fixed_name == String(), nullptr);
+				ERR_FAIL_COND_V(fixed_node.first == String(), nullptr);
 
 				if (shapes.size()) {
-					StaticBody3D *col = memnew(StaticBody3D);
-					col->set_transform(mi->get_transform());
-					col->set_name(fixed_name);
-					p_node->replace_by(col);
-					memdelete(p_node);
-					p_node = col;
-
-					int idx = 0;
-					for (List<Ref<Shape3D>>::Element *E = shapes.front(); E; E = E->next()) {
-						CollisionShape3D *cshape = memnew(CollisionShape3D);
-						cshape->set_shape(E->get());
-						col->add_child(cshape);
-
-						cshape->set_name("shape" + itos(idx));
-						cshape->set_owner(col->get_owner());
-						idx++;
+					if (fixed_node.second == NODE_HINT_COLLISION_ONLY || fixed_node.second == NODE_HINT_CONVEX_COLLISION_ONLY) {
+						StaticBody3D *static_body = memnew(StaticBody3D);
+						static_body->set_transform(mesh_instance->get_transform());
+						static_body->set_name(fixed_node.first);
+						p_node->replace_by(static_body);
+						memdelete(p_node);
+						p_node = static_body;
+
+						_add_shapes(static_body, shapes);
+					} else {
+						_add_shapes(p_root, shapes, mesh_instance->get_transform(), fixed_node.first);
+						memdelete(p_node);
+						p_node = nullptr;
 					}
 				}
 			}
 
-		} else if (p_node->has_meta("empty_draw_type")) {
-			String empty_draw_type = String(p_node->get_meta("empty_draw_type"));
-			StaticBody3D *sb = memnew(StaticBody3D);
-			sb->set_name(_fixstr(name, "colonly"));
-			Object::cast_to<Node3D>(sb)->set_transform(Object::cast_to<Node3D>(p_node)->get_transform());
-			p_node->replace_by(sb);
-			memdelete(p_node);
-			p_node = nullptr;
-			CollisionShape3D *colshape = memnew(CollisionShape3D);
-			if (empty_draw_type == "CUBE") {
-				BoxShape3D *boxShape = memnew(BoxShape3D);
-				boxShape->set_extents(Vector3(1, 1, 1));
-				colshape->set_shape(boxShape);
-				colshape->set_name("BoxShape3D");
-			} else if (empty_draw_type == "SINGLE_ARROW") {
-				RayShape3D *rayShape = memnew(RayShape3D);
-				rayShape->set_length(1);
-				colshape->set_shape(rayShape);
-				colshape->set_name("RayShape3D");
-				Object::cast_to<Node3D>(sb)->rotate_x(Math_PI / 2);
-			} else if (empty_draw_type == "IMAGE") {
-				WorldMarginShape3D *world_margin_shape = memnew(WorldMarginShape3D);
-				colshape->set_shape(world_margin_shape);
-				colshape->set_name("WorldMarginShape3D");
+		} else if (CollisionShape3D *colshape = _shape_from_empty_meta(p_node); colshape) {
+			if (fixed_node.second == NODE_HINT_COLLISION_ONLY || fixed_node.second == NODE_HINT_CONVEX_COLLISION_ONLY) {
+				StaticBody3D *static_body = memnew(StaticBody3D);
+				static_body->set_name(fixed_node.first);
+				p_node->replace_by(static_body);
+				memdelete(p_node);
```

Comment:
Is this correct?

You replace the pointer by the static body. I haven't looked into this enough.

Reply:
I tend to use queue delete, but the definition of next engine tick before saving is a bit confusing.

Reply:
Not sure, I'm new here, but this is how it was https://github.com/godotengine/godot/blob/59fdfc77e8f3d455ff6c86c73c27b04b022843c8/editor/import/resource_importer_scene.cpp#L397. I just added a new condition for shapes, removed redundant `set_name` and reworked parsing hint (parsing was done on every check!).